### DISCUSSION TITLE
How to merge lines in text file

### DISCUSSION THREAD

**Discussion by Shahin-rmz** [saulpw/visidata#1220] [Q&A]
I have a .txt file, which can be opened without problem.
the problem is, each line will be shown as one separate row.
I may want to merge the rows together and see multiple lines as *one row*. But haven't seen such feature.
Also tried ```wrap``` but that was something else.
Thanks for considering this feature.

**Comment by geekscrapy** [comment]
Great idea, I sometimes have this usecase when working with log files which contain mulitline exceptions (e.g. java). I'd like to be able to define a starting match, and an ending match, anything in-between is merged.

**Comment by saulpw** [comment]
Hi @Shahin-rmz, you can use create a column that indicates which 'line' each line is supposed to be on, and `+list` on the text column, and then Shift+F (freq) on the new column to turn the text into a list of lines per group.  Then you can use `='\n'.join(text_list)` to create a column with the lines joined together.

I would need actual example data and the desired result to guide you further.

**Comment by Shahin-rmz** [comment]
Hi @saulpw , again another question regarding merging.
I want to `join` to columns in a csv file.
I can add them together with `=col1+col2` but I think it is not optimal. looking for `join` method.
is there any?
Thanks

**Reply by ghost** [reply]
Do you want the equivalent of Python's `str.join` function? Or do you want the equivalent of a SQL table join?

**Reply by Shahin-rmz** [reply]
I man, more of sql join, where we can do right join, left join, inner join etc.

**Reply by aborruso** [reply]
> I man, more of sql join, where we can do right join, left join, inner join etc.

VisiData has a good online documentation https://www.visidata.org/docs/join/

**Reply by Shahin-rmz** [reply]
It's between sheets, I need for columns

**Reply by ghost** [reply]
Could you give an example of what the columns look like before and after you join them?
It's not exactly clear what you're trying to do.

**Comment by Shahin-rmz** [comment]
sure.
|col A| col B| col A join B|
|------|------|-----------|
|1     |     | 1 |
| | 2 | 2|
|3|3|3

**Reply by Shahin-rmz** [reply]
also with left and right and full join

**Reply by ghost** [reply]
Are you trying to do the equivalent of `coalesce` in SQL, which picks the first non-null value? ([example](https://www.w3schools.com/SQL/func_sqlserver_coalesce.asp))

**Reply by ghost** [reply]
I'm not quite sure what you mean by left, right and full join for this? If you were doing an inner join between col A and col B, you would have just gotten null for the first two rows, and 3 for the last row.

**Reply by Shahin-rmz** [reply]
@davidwales `coalesce` is pretty much neat.
it sums up the idea pretty good.

**Reply by ghost** [reply]
You can use a Python expression like this:

```python
=A or B
```

This will take the first non-null value (from left to right)

|A   |B   |A or B  |
|----|----|--------|
|1   |2   |1       |
|    |3   |3       |
|4   |    |4       |
|    |    |\(None\)|

If you want anything fancier, you'll likely need to create your own function and add it to your `~/.visidatarc`, or create a plugin.

**Reply by anjakefala** [reply]
Thanks for answering @davidwales! That sounds about right.

**Comment by Shahin-rmz** [comment]
@davidwales you are awesome.
sorry if superficial question.
But very helpful of you.
